fv3core/stencils/delnflux.py

```diff
@@ -100,93 +410,256 @@ def compute_delnflux_no_sg(
+        nk: Number of k-levels to compute on (in)
+        mass: Mass to weight the diffusive flux by (in)
+    """
+    grid = spec.grid
+    i1 = grid.is_ - 1 - nord
+    i2 = grid.ie + 1 + nord
+    j1 = grid.js - 1 - nord
+    j2 = grid.je + 1 + nord
+    if nk is None:
+        nk = grid.npz - kstart
+    origin_d2 = (i1, j1, kstart)
+    domain_d2 = (i2 - i1 + 1, j2 - j1 + 1, nk)
+    fxy_stencil = gtstencil(definition=fxy_order, externals={"nord": nord})
+    if mass is None:
+        d2_damp(q, d2, damp_c, origin=origin_d2, domain=domain_d2)
+    else:
+        d2 = copy(q, origin=origin_d2, domain=domain_d2)
+
+    fxy_stencil(
+        d2,
+        grid.del6_u,
+        grid.del6_v,
+        fx,
+        fy,
+        order=1,
+        origin=(grid.isd, grid.jsd, kstart),
+        domain=(grid.nid, grid.njd, nk),
+    )
+
+    if nord > 0:
+        for n in range(nord):
+            nt = nord - 1 - n
+            looped_stencil = gtstencil(
+                definition=higher_order_compute, externals={"nt": nt}
+            )
+            looped_stencil(
+                fx,
+                fy,
+                grid.rarea,
+                d2,
+                grid.del6_u,
+                grid.del6_v,
+                order=2 + n,
+                origin=(grid.isd, grid.jsd, kstart),
+                domain=(grid.nid, grid.njd, nk),
+            )
+
+
+def compute_no_sg_unroll(
+    q: FloatField,
+    fx: FloatField,
+    fy: FloatField,
+    nord: int,
+    damp_c: float,
+    d2: FloatField,
+    kstart: Optional[int] = 0,
+    nk: Optional[int] = None,
+    mass: Optional["FloatField"] = None,
+):
+    """
+    Calculate deln-fluxes by unrolling the loop over nord.
+    We'll apply diffusion later.
+    Args:
+        q: Tracer field (in)
+        fx: x-flux on A-grid (inout)
+        fy: y-flux on A-grid (inout)
+        nord: Order of divergence damping (in)
+        damp_c: damping coefficient (in)
+        d2: A damped copy of the q field (in)
+        kstart: k-level to begin computing on (in)
+        nk: Number of k-levels to compute on (in)
+        mass: Mass to weight the diffusive flux by (in)
+    """
+    grid = spec.grid
+    i1 = grid.is_ - 1 - nord
+    i2 = grid.ie + 1 + nord
+    j1 = grid.js - 1 - nord
+    j2 = grid.je + 1 + nord
+    if nk is None:
+        nk = grid.npz - kstart
+    origin_d2 = (i1, j1, kstart)
+    domain_d2 = (i2 - i1 + 1, j2 - j1 + 1, nk)
+    if mass is None:
+        d2_damp(q, d2, damp_c, origin=origin_d2, domain=domain_d2)
+    else:
+        d2 = copy(q, origin=origin_d2, domain=domain_d2)
+
+    if nord > 0:
+        if nord == 3:  # 2, 1, 0
+            higher_order_compute_unroll3(
+                fx,
+                fy,
+                grid.rarea,
+                d2,
+                grid.del6_u,
+                grid.del6_v,
+                origin=(grid.isd, grid.jsd, kstart),
+                domain=(grid.nid, grid.njd, nk),
+            )
```

Comment:
ah this is more complicated than divergence damping, because it looks like delnflux might in fact get called with different nords, while divergence damping always has nord = namelist value or 0.

Reply:
You may also then get to reuse some of the functions in the other unroll function

Reply:
Are the `nord` values known at compile time? Maybe something we could unroll in the backend later?